mocks/timestamp.js

```diff
@@ -19,34 +19,34 @@ class Timestamp {
   }
 
   toDate() {
-    const d = new Date(0);
-    d.setTime(this.seconds * 1000);
-    d.setMilliseconds(this.nanoseconds / 1000000);
-    return mockTimestampToDate(...arguments) || d;
+    return mockTimestampToDate(...arguments) || new Date(this._toMillis());
   }
 
   toMillis() {
-    const d = new Date(0);
-    d.setSeconds(this.seconds);
-    d.setMilliseconds(this.nanoseconds / 1000000);
-    return mockTimestampToMillis(...arguments) || d.getMilliseconds();
+    return mockTimestampToMillis(...arguments) || this._toMillis();
   }
 
   valueOf() {
     return JSON.stringify(this.toMillis());
   }
 
   static fromDate(date) {
-    return (
-      mockTimestampFromDate(...arguments) ||
-      new Timestamp(date.getTime() * 0.001, date.getMilliseconds() * 1000000)
-    );
+    return mockTimestampFromDate(...arguments) || Timestamp._fromMillis(date.getTime());
   }
 
   static fromMillis(millis) {
-    const d = new Date(0);
-    d.setMilliseconds(millis);
-    return mockTimestampFromMillis(...arguments) || Timestamp.fromDate(d);
```

Comment:
part of extracting the logic into helper functions was to fix this issue: the original resulted in incorrectly calling `mockTimestampFromDate`